src/storage/mod.rs

```diff
@@ -6574,18 +6594,110 @@ mod tests {
+            block_on(storage.batch_get_command(
+                vec![req1.clone(), req2],
+                vec![1, 2],
+                consumer.clone(),
+                Instant::now(),
+            ))
+            .unwrap();
+            consumer.take_data()
+        };
+        let res = batch_get_command(req2.clone());
         assert!(res[0].is_ok());
         let key_error = extract_key_error(res[1].as_ref().unwrap_err());
         assert_eq!(key_error.get_locked().get_key(), b"key");
+        // Ignore memory locks in resolved or committed locks.
+        req2.mut_context().set_resolved_locks(vec![10]);
+        let res = batch_get_command(req2.clone());
+        assert!(res[0].is_ok());
+        assert!(res[1].is_ok());
+        req2.mut_context().take_resolved_locks();
+    }
+
+    #[test]
+    fn test_read_access_locks() {
+        let storage = TestStorageBuilder::new(DummyLockManager {}, false)
+            .build()
+            .unwrap();
+
+        let (k1, v1) = (b"k1".to_vec(), b"v1".to_vec());
+        let (k2, v2) = (b"k2".to_vec(), b"v2".to_vec());
+        let (tx, rx) = channel();
+        storage
+            .sched_txn_command(
+                commands::Prewrite::with_defaults(
+                    vec![
+                        Mutation::Put((Key::from_raw(&k1), v1.clone())),
+                        Mutation::Put((Key::from_raw(&k2), v2.clone())),
+                    ],
+                    k1.clone(),
+                    100.into(),
+                ),
+                expect_ok_callback(tx, 0),
+            )
+            .unwrap();
+        rx.recv().unwrap();
+
+        let mut ctx = Context::default();
+        ctx.set_isolation_level(IsolationLevel::Si);
+        ctx.set_committed_locks(vec![100]);
+        // get
+        assert_eq!(
+            block_on(storage.get(ctx.clone(), k1.clone(), 110.into()))
+                .unwrap()
+                .0,
+            Some(v1.clone())
+        );
+        // batch get
+        let res =
+            block_on(storage.batch_get(ctx.clone(), vec![k1.clone(), k2.clone()], 110.into()))
+                .unwrap()
+                .0;
+        if res[0].as_ref().unwrap().0 == k1 {
+            assert_eq!(&res[0].as_ref().unwrap().1, &v1);
+            assert_eq!(&res[1].as_ref().unwrap().1, &v2);
+        } else {
+            assert_eq!(&res[0].as_ref().unwrap().1, &v2);
+            assert_eq!(&res[1].as_ref().unwrap().1, &v1);
+        }
+        // batch get commands
+        let mut req = GetRequest::default();
+        req.set_context(ctx.clone());
+        req.set_key(k1.clone());
+        req.set_version(110);
+        let consumer = GetConsumer::new();
+        block_on(storage.batch_get_command(vec![req], vec![1], consumer.clone(), Instant::now()))
+            .unwrap();
+        let res = consumer.take_data();
+        assert_eq!(res.len(), 1);
+        assert_eq!(res[0].as_ref().unwrap(), &Some(v1.clone()));
+        // scan
+        for desc in &[false] {
```

Comment:
May also test reverse scan?

Reply:
Yes. Seems I forgot to add it.

Reply:
I just noticed that the `start_key` should be greater than `end_key` in reverse scan, but checking memory locks still uses [`start_key`, `end_key`). /cc @sticnarf